File: src/usage.mjs
/**
 * Usage snapshot for model-scoped weekly buckets.
 *
 * Claude Code's statusline stdin already carries the 5-hour and weekly windows,
 * so this module exists for the one thing stdin does NOT expose: per-model
 * weekly quotas (`kind: "weekly_scoped"`, e.g. Fable). It reads them from
 * `GET api.anthropic.com/api/oauth/usage` using the OAuth token Claude Code
 * already stores.
 *
 * Design rules:
 *  - The render path never blocks on the network. It reads a cached snapshot
 *    and, when that snapshot is stale, spawns a detached refresh for the *next*
 *    frame. A cold cache costs the scoped buckets one frame, nothing else.
 *  - We never refresh the OAuth token ourselves. That is Claude Code's job, and
 *    racing it risks invalidating the credentials the editor is using. An
 *    expired token simply means no scoped buckets until Claude Code renews it.
 */

import { execFileSync } from "node:child_process";
import { createHash } from "node:crypto";
import { existsSync, readFileSync, renameSync, writeFileSync, mkdirSync, statSync, rmSync } from "node:fs";
import https from "node:https";
import { join } from "node:path";

import { cacheDir, claudeConfigDir, ensureCacheDir } from "./config.mjs";

const USAGE_HOST = "api.anthropic.com";
const USAGE_PATH = "/api/oauth/usage";
const OAUTH_BETA = "oauth-2025-04-20";
const REQUEST_TIMEOUT_MS = 6000;
const LOCK_STALE_MS = 30_000;

export function usageCachePath(env = process.env) {
  return join(cacheDir(env), "usage.json");
}

function lockPath(env = process.env) {
  return join(cacheDir(env), "usage.lock");
}

/**
 * Read the OAuth access token.
 *
 * Linux/WSL keep it in `~/.claude/.credentials.json`; macOS keeps it in the
 * login Keychain under a config-dir-scoped account name.
 */
export function readAccessToken(env = process.env) {
  const filePath = join(claudeConfigDir(env), ".credentials.json");
  if (existsSync(filePath)) {
    const token = tokenFromJson(safeReadJson(filePath));
    if (token) return token;
  }
  if (process.platform === "darwin") {
    const token = tokenFromJson(readKeychainCredentials(env));
    if (token) return token;
  }
  return null;
}

function safeReadJson(path) {
  try {
    return JSON.parse(readFileSync(path, "utf8"));
  } catch {
    return null;
  }
}

function tokenFromJson(json) {
  if (!json || typeof json !== "object") return null;
  const oauth = json.claudeAiOauth ?? json;
  const token = oauth?.accessToken ?? oauth?.access_token;
  if (typeof token !== "string" || token === "") return null;
  const expiresAt = oauth?.expiresAt ?? oauth?.expires_at;
  if (typeof expiresAt === "number" && Number.isFinite(expiresAt) && expiresAt <= Date.now()) {
    return null; // expired — Claude Code will renew it; we must not.
  }
  return token;
}

function readKeychainCredentials(env) {
  // Claude Code scopes the Keychain item by config dir when it is non-default.
  const configDir = claudeConfigDir(env);
  const defaultDir = join(env.HOME ?? "", ".claude");
  const service =
    configDir === defaultDir
      ? "Claude Code-credentials"
      : `Claude Code-credentials-${createHash("sha256").update(configDir).digest("hex").slice(0, 8)}`;
  try {
    const raw = execFileSync("security", ["find-generic-password", "-s", service, "-w"], {
      encoding: "utf8",
      stdio: ["ignore", "pipe", "ignore"],
      timeout: 3000,
    });
    return JSON.parse(raw);
  } catch {
    return null;
  }
}

/** One HTTPS GET against the usage endpoint. Resolves to the parsed body. */
function fetchUsage(token) {
  return new Promise((resolve, reject) => {
    const req = https.request(
      {
        hostname: USAGE_HOST,
        path: USAGE_PATH,
        method: "GET",
        headers: {
          Authorization: `Bearer ${token}`,
          "anthropic-beta": OAUTH_BETA,
          Accept: "application/json",
        },
        timeout: REQUEST_TIMEOUT_MS,
      },
      (res) => {
        let body = "";
        res.setEncoding("utf8");
        res.on("data", (chunk) => {
          body += chunk;
          if (body.length > 1_000_000) req.destroy(new Error("usage response too large"));
        });
        res.on("end", () => {
          if (res.statusCode !== 200) {
            reject(new Error(`usage endpoint returned ${res.statusCode}`));
            return;
          }
          try {
            resolve(JSON.parse(body));
          } catch (error) {
            reject(error);
          }
        });
      },
    );
    req.on("timeout", () => req.destroy(new Error("usage request timed out")));
    req.on("error", reject);
    req.end();
  });
}

function clampPercent(value) {
  if (!Number.isFinite(value)) return null;
  return Math.max(0, Math.min(100, Math.round(value)));
}

/**
 * Pull model-scoped weekly buckets out of the response.
 *
 * The `limits[]` array is the forward-compatible shape: a new model tier shows
 * up as another `weekly_scoped` entry with its own `scope.model.display_name`,
 * so it renders without a release here. Entries are de-duplicated by display
 * name, preferring the active one.
 */
export function parseScopedBuckets(response) {
  const limits = response?.limits;
  if (!Array.isArray(limits)) return [];
  const byName = new Map();
  for (const entry of limits) {
    if (!entry || typeof entry !== "object" || entry.kind !== "weekly_scoped") continue;
    const percent = clampPercent(entry.percent);
    if (percent === null) continue;
    const name = entry.scope?.model?.display_name;
    if (typeof name !== "string" || name.trim() === "") continue;
    const label = name.trim();
    const key = label.toLowerCase();
    const isActive = entry.is_active === true;
    if (!byName.has(key) || (isActive && !byName.get(key).isActive)) {
      byName.set(key, {
        id: entry.scope?.model?.id ?? key,
        label,
        percent,
        resetsAt: entry.resets_at ?? null,
        severity: typeof entry.severity === "string" ? entry.severity : null,
        isActive,
      });
    }
  }
  return [...byName.values()];
}

export function parseUsage(response) {
  return {
    scoped: parseScopedBuckets(response),
    fetchedAt: Date.now(),
  };
}

export function readCachedUsage(env = process.env) {
  const cached = safeReadJson(usageCachePath(env));
  if (!cached || typeof cached !== "object" || !Array.isArray(cached.scoped)) return null;
  return cached;
}

export function isStale(cached, ttlSeconds) {
  if (!cached?.fetchedAt) return true;
  return Date.now() - cached.fetchedAt > ttlSeconds * 1000;
}

function writeCache(payload, env = process.env) {
  ensureCacheDir(env);
  const target = usageCachePath(env);
  const tmp = `${target}.${process.pid}.tmp`;
  writeFileSync(tmp, JSON.stringify(payload), { mode: 0o600 });
  renameSync(tmp, target);
}

/** Single-writer guard so concurrent panes do not all hit the API at once. */
function acquireLock(env = process.env) {
  ensureCacheDir(env);
  const path = lockPath(env);
  try {
    mkdirSync(path);
    return true;
  } catch {
    try {
      if (Date.now() - statSync(path).mtimeMs > LOCK_STALE_MS) {
        rmSync(path, { recursive: true, force: true });
        mkdirSync(path);
        return true;
      }
    } catch {
      /* lost the race to another refresher */
    }
    return false;
  }
}

/**
 * Claim the right to refresh, for the caller to hand to a spawned child.
 *
 * The parent takes the lock *before* spawning: a child needs ~20ms to boot, and
 * during that window every further frame would otherwise see an unlocked cache
 * and spawn its own refresher. Taking it here makes the claim visible
 * immediately, so a burst of frames produces exactly one fetch.
 *
 * The caller MUST pass ownership to a process that releases it, or call
 * `abandonRefresh()` — otherwise the lock clears on its own after
 * `LOCK_STALE_MS`.
 */
export function claimRefresh(env = process.env) {
  return acquireLock(env);
}

export function abandonRefresh(env = process.env) {
  releaseLock(env);
}

function releaseLock(env = process.env) {
  try {
    rmSync(lockPath(env), { recursive: true, force: true });
  } catch {
    /* best effort */
  }
}

/**
 * Fetch and cache a fresh snapshot. Runs in the detached refresh process, so
 * every failure is swallowed — a stale bar beats a broken statusline.
 */
export async function refreshUsage(env = process.env, { lockHeld = false } = {}) {
  if (!lockHeld && !acquireLock(env)) return null;
  try {
    const token = readAccessToken(env);
    if (!token) {
      writeCache({ scoped: [], fetchedAt: Date.now(), error: "no-token" }, env);
      return null;
    }
    const parsed = parseUsage(await fetchUsage(token));
    writeCache(parsed, env);
    return parsed;
  } catch (error) {
    // Keep the previous buckets if we have them; only the timestamp moves, so
    // the next attempt waits a full TTL instead of hammering the endpoint.
    const previous = readCachedUsage(env);
    writeCache(
      { scoped: previous?.scoped ?? [], fetchedAt: Date.now(), error: error?.message ?? "fetch failed" },
      env,
    );
    return null;
  } finally {
    releaseLock(env);
  }
}

File: usage.go
package main

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"math"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
	"time"
)

// The render path never blocks on the network. It reads a cached snapshot and,
// when that is stale, spawns a detached refresh for the next frame. The OAuth
// token is never refreshed here; that is Claude Code's job, and racing it risks
// invalidating the credentials the editor is using.

const (
	usageURL        = "https://api.anthropic.com/api/oauth/usage"
	oauthBeta       = "oauth-2025-04-20"
	usageTimeout    = 6 * time.Second
	usageBodyLimit  = 1_000_000
	lockStale       = 30 * time.Second
	keychainTimeout = 3 * time.Second
	keychainService = "Claude Code-credentials"
)

// Bucket is one model-scoped weekly quota. JSON tags match the cache file the
// Node version wrote, so an existing ~/.cache/claude-readout/usage.json keeps
// working after the upgrade.
type Bucket struct {
	ID       string    `json:"id"`
	Label    string    `json:"label"`
	Percent  int       `json:"percent"`
	ResetsAt Timestamp `json:"resetsAt"`
	Severity *string   `json:"severity"`
	IsActive bool      `json:"isActive"`
}

// UsageCache is the on-disk snapshot. FetchedAt is epoch milliseconds.
type UsageCache struct {
	Scoped    []Bucket `json:"scoped"`
	FetchedAt int64    `json:"fetchedAt"`
	Error     string   `json:"error,omitempty"`
}

func (e Env) usageCachePath() string { return filepath.Join(e.cacheDir(), "usage.json") }
func (e Env) lockPath() string       { return filepath.Join(e.cacheDir(), "usage.lock") }

// readAccessToken reads Claude Code's OAuth access token: from
// <claudeConfigDir>/.credentials.json, else on macOS from the login Keychain.
// The Keychain service name is the one Claude Code itself uses, which it
// suffixes with the first 8 hex chars of the SHA-256 of the config dir when
// CLAUDE_CONFIG_DIR points somewhere other than $HOME/.claude. Both sources
// hold {"claudeAiOauth": {"accessToken", "expiresAt"}}; a bare {"accessToken"}
// or snake_case keys also work. An expired token reads as absent.
func readAccessToken(env Env, now time.Time) (string, bool) {
	credentials, err := os.ReadFile(filepath.Join(env.claudeConfigDir(), ".credentials.json"))
	if err == nil {
		if token, ok := accessTokenFromJSON(credentials, now); ok {
			return token, true
		}
	}
	if runtime.GOOS != "darwin" {
		return "", false
	}

	service := keychainService
	configDir := env.claudeConfigDir()
	if configDir != filepath.Join(env.Home, ".claude") {
		sum := sha256.Sum256([]byte(configDir))
		service += "-" + hex.EncodeToString(sum[:])[:8]
	}
	ctx, cancel := context.WithTimeout(context.Background(), keychainTimeout)
	defer cancel()
	cmd := exec.CommandContext(ctx, "security", "find-generic-password", "-s", service, "-w")
	cmd.Stderr = io.Discard
	raw, err := cmd.Output()
	if err != nil {
		return "", false
	}
	return accessTokenFromJSON(raw, now)
}

func accessTokenFromJSON(data []byte, now time.Time) (string, bool) {
	var root map[string]json.RawMessage
	if err := json.Unmarshal(data, &root); err != nil || root == nil {
		return "", false
	}
	oauth := root
	if raw, ok := root["claudeAiOauth"]; ok && !isJSONNull(raw) {
		var nested map[string]json.RawMessage
		if err := json.Unmarshal(raw, &nested); err != nil || nested == nil {
			return "", false
		}
		oauth = nested
	}

	tokenRaw, ok := oauth["accessToken"]
	if !ok || isJSONNull(tokenRaw) {
		tokenRaw, ok = oauth["access_token"]
	}
	if !ok {
		return "", false
	}
	var token string
	if err := json.Unmarshal(tokenRaw, &token); err != nil || token == "" {
		return "", false
	}

	expiresRaw, ok := oauth["expiresAt"]
	if !ok || isJSONNull(expiresRaw) {
		expiresRaw, ok = oauth["expires_at"]
	}
	if ok {
		var expires any
		if json.Unmarshal(expiresRaw, &expires) == nil {
			if value, numeric := expires.(float64); numeric && !math.IsNaN(value) && !math.IsInf(value, 0) && value <= float64(now.UnixMilli()) {
				return "", false
			}
		}
	}
	return token, true
}

func isJSONNull(raw json.RawMessage) bool {
	return strings.TrimSpace(string(raw)) == "null"
}

// parseScopedBuckets pulls kind == "weekly_scoped" entries out of the usage
// response's limits[] array. Percent is rounded and clamped to 0..100; an entry
// with a non-numeric percent or a blank scope.model.display_name is skipped.
// Entries are de-duplicated by lower-cased display name, an active entry
// replacing an inactive one. ID falls back to the lower-cased name. Order is
// first appearance.
func parseScopedBuckets(response []byte) []Bucket {
	var root map[string]json.RawMessage
	if err := json.Unmarshal(response, &root); err != nil || root == nil {
		return []Bucket{}
	}
	var limits []json.RawMessage
	if err := json.Unmarshal(root["limits"], &limits); err != nil {
		return []Bucket{}
	}

	buckets := make([]Bucket, 0)
	positions := make(map[string]int)
	for _, raw := range limits {
		var entry map[string]json.RawMessage
		if err := json.Unmarshal(raw, &entry); err != nil || entry == nil {
			continue
		}
		var kind string
		if json.Unmarshal(entry["kind"], &kind) != nil || kind != "weekly_scoped" {
			continue
		}
		var percentValue any
		if json.Unmarshal(entry["percent"], &percentValue) != nil {
			continue
		}
		percentNumber, ok := percentValue.(float64)
		if !ok || math.IsNaN(percentNumber) || math.IsInf(percentNumber, 0) {
			continue
		}

		var scope map[string]json.RawMessage
		var model map[string]json.RawMessage
		if json.Unmarshal(entry["scope"], &scope) != nil || json.Unmarshal(scope["model"], &model) != nil {
			continue
		}
		var displayName string
		if json.Unmarshal(model["display_name"], &displayName) != nil || strings.TrimSpace(displayName) == "" {
			continue
		}
		label := strings.TrimSpace(displayName)
		key := strings.ToLower(label)
		id := ""
		if json.Unmarshal(model["id"], &id) != nil || id == "" {
			id = key
		}
		isActive := false
		_ = json.Unmarshal(entry["is_active"], &isActive)
		var severity *string
		var severityAny any
		if json.Unmarshal(entry["severity"], &severityAny) == nil {
			if severityValue, ok := severityAny.(string); ok {
				severity = &severityValue
			}
		}
		var resetsAt Timestamp
		_ = json.Unmarshal(entry["resets_at"], &resetsAt)
		bucket := Bucket{
			ID:       id,
			Label:    label,
			Percent:  int(math.Max(0, math.Min(100, jsRound(percentNumber)))),
			ResetsAt: resetsAt,
			Severity: severity,
			IsActive: isActive,
		}
		if position, exists := positions[key]; !exists {
			positions[key] = len(buckets)
			buckets = append(buckets, bucket)
		} else if isActive && !buckets[position].IsActive {
			buckets[position] = bucket
		}
	}
	return buckets
}

// readCachedUsage returns the snapshot, or nil when missing or malformed
// (a snapshot without a "scoped" array counts as malformed).
func readCachedUsage(env Env) *UsageCache {
	data, err := os.ReadFile(env.usageCachePath())
	if err != nil {
		return nil
	}
	var cached UsageCache
	if json.Unmarshal(data, &cached) != nil || cached.Scoped == nil {
		return nil
	}
	return &cached
}

// isStale is true for a nil or never-fetched snapshot, or one older than the TTL.
func isStale(cached *UsageCache, ttlSeconds float64, now time.Time) bool {
	if cached == nil || cached.FetchedAt == 0 {
		return true
	}
	return float64(now.UnixMilli()-cached.FetchedAt) > ttlSeconds*1000
}

// claimRefresh takes the single-writer lock (a directory, created with mkdir;
// one older than lockStale is removed and retaken). It returns false when
// another refresher holds it. The caller must hand ownership to a process that
// releases it, or call abandonRefresh.
func claimRefresh(env Env) bool {
	env.ensureCacheDir()
	path := env.lockPath()
	if err := os.Mkdir(path, 0o777); err == nil {
		return true
	}
	info, err := os.Stat(path)
	if err != nil || time.Since(info.ModTime()) <= lockStale {
		return false
	}
	if os.RemoveAll(path) != nil {
		return false
	}
	return os.Mkdir(path, 0o777) == nil
}

func abandonRefresh(env Env) {
	_ = os.RemoveAll(env.lockPath())
}

// fetchUsage GETs the usage endpoint with the bearer token. Redirects are not
// followed, so the token cannot be handed to another host.
func fetchUsage(token string) ([]byte, error) {
	req, err := http.NewRequest(http.MethodGet, usageURL, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Bearer "+token)
	req.Header.Set("anthropic-beta", oauthBeta)
	req.Header.Set("Accept", "application/json")
	client := http.Client{
		Timeout: usageTimeout,
		CheckRedirect: func(*http.Request, []*http.Request) error {
			return http.ErrUseLastResponse
		},
	}
	response, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer response.Body.Close()
	body, err := io.ReadAll(io.LimitReader(response.Body, usageBodyLimit+1))
	if err != nil {
		return nil, err
	}
	if len(body) > usageBodyLimit {
		return nil, fmt.Errorf("usage response too large")
	}
	if response.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("usage endpoint returned %d", response.StatusCode)
	}
	return body, nil
}

// refreshUsage fetches and caches a snapshot. It runs in the detached child, so
// every failure is swallowed: a stale bar beats a broken statusline. On a
// failure after the token was found it keeps the previous buckets and moves
// FetchedAt to now, so the next attempt waits a full TTL instead of hammering
// a broken endpoint every frame.
func refreshUsage(env Env, lockHeld bool, fetch func(token string) ([]byte, error), now time.Time) *UsageCache {
	if !lockHeld && !claimRefresh(env) {
		return nil
	}
	defer abandonRefresh(env)

	writeCache := func(cache UsageCache) {
		env.ensureCacheDir()
		data, err := json.Marshal(cache)
		if err != nil {
			return
		}
		target := env.usageCachePath()
		tmp := fmt.Sprintf("%s.%d.tmp", target, os.Getpid())
		if err := os.WriteFile(tmp, data, 0o600); err != nil {
			return
		}
		_ = os.Rename(tmp, target)
	}

	token, ok := readAccessToken(env, now)
	if !ok {
		writeCache(UsageCache{Scoped: []Bucket{}, FetchedAt: now.UnixMilli(), Error: "no-token"})
		return nil
	}
	body, err := fetch(token)
	if err == nil && !json.Valid(body) {
		err = errors.New("usage response is not JSON")
	}
	if err != nil {
		previous := readCachedUsage(env)
		scoped := []Bucket{}
		if previous != nil {
			scoped = previous.Scoped
		}
		writeCache(UsageCache{Scoped: scoped, FetchedAt: now.UnixMilli(), Error: err.Error()})
		return nil
	}

	cache := UsageCache{Scoped: parseScopedBuckets(body), FetchedAt: now.UnixMilli()}
	writeCache(cache)
	return &cache
}
